easybuild/easyconfigs/m/M4/M4-1.4.18.eb

```diff
@@ -16,12 +16,20 @@ toolchain = {'name': 'dummy', 'version': 'dummy'}
 
 source_urls = [GNU_SOURCE]
 sources = [SOURCELOWER_TAR_GZ]
-checksums = ['ab2633921a5cd38e48797bf5521ad259bdc4b979078034a3b790d7fec5493fab']
+
+# fix problems occuring from changes in glibc 2.28
+# https://github.com/coreutils/gnulib/commit/4af4a4a71827c0bc5e0ec67af23edef4f15cee8e
+patches = ['M4_glibc_2.28.patch']
```

Comment:
@robertschade Do you mind renaming the patch file to something like `M4-1.4.18_glibc-2.28.patch`?

We try to keep the software version for which the patch was created in the patch filename, that makes it easier w.r.t. maintenance of patches.

Reply:
Yes, I would be happy to do the suggested changes. I will update the PR later today.

Reply:
@boegel One problem is that even M4-1.4.18-GCCcore-8.2.0.eb depend on Bison 3.0.4 via flex 2.6.4. However Bison 3.0.4 fails on glibc-2.28 with a similar issue like M4-1.4.18 (even if M4-1.4.17 is pachted). Bison 3.0.5 compiles under glibc 2.28.
There are now two options:
1. patch Bison 3.0.4 to work under glibc 2.28
2. change all requirements of Bison 3.0.4 to Bison 3.0.5

Which one do you prefer?

Reply:
@robertschade Where exactly do you see this dependency of `M4` on `flex -> Bison`?

binutils-2.31.1.eb only includes `binutils` as a (build) dependency, plus the `GCCcore` toolchain.

`binutils` does depend on both `flex` and `Bison`, but `binutils-2.31.1.eb` (which is the one used as build dep for `binutils-2.31.1.eb`) already depends on `Bison` 3.0.5`?

In general, we prefer not to bump versions of dependencies, but in the case of *build* only dependencies we can make an exception. In the long run, patching Bison 3.0.4 may be worth the effort though (preferably in a separate PR)... How big of an effort is that?

Reply:
@boegel  The dependency chains are:

GCCcore-8.2.0.eb -> binutils-2.31.1.eb -> flex-2.6.4.eb -> Bison-3.0.4.eb -> M4-1.4.17.eb

GCCcore-8.2.0.eb -> binutils-2.31.1.eb -> M4-1.4.18.eb

So that one problem is that M4-1.4.17.eb and M4-1.4.18.eb need to be patched and installed. The other problem is that even with patched M4-1.4.17.eb and M4-1.4.18.eb the compilation of Bison-3.0.4.eb fails.

IMHO, the cleanest solution would be to bump the dependency of  flex-2.6.4.eb up to Bison-3.0.5.eb. What do you think?

```eb ~/tmp/easybuild_M4/prefix/software/EasyBuild/3.8.0/lib/python2.7/site-packages/easybuild_easyconfigs-3.8.0-py2.7.egg/easybuild/easyconfigs/g/GCCcore/GCCcore-8.2.0.eb  -k -D
== temporary log file in case of crash /tmp/eb-aTAklR/easybuild-dHqsp3.log
Dry run: printing build status of easyconfigs and dependencies
CFGS=/home/rschade/tmp/easybuild_M4/prefix/software/EasyBuild/3.8.0/lib/python2.7/site-packages/easybuild_easyconfigs-3.8.0-py2.7.egg/easybuild/easyconfigs
 * [ ] $CFGS/m/M4/M4-1.4.18.eb (module: M4/1.4.18)
 * [ ] $CFGS/b/Bison/Bison-3.0.5.eb (module: Bison/3.0.5)
 * [ ] $CFGS/z/zlib/zlib-1.2.11.eb (module: zlib/1.2.11)
 * [ ] $CFGS/h/help2man/help2man-1.47.4.eb (module: help2man/1.47.4)
 * [ ] $CFGS/m/M4/M4-1.4.17.eb (module: M4/1.4.17)
 * [ ] $CFGS/b/Bison/Bison-3.0.4.eb (module: Bison/3.0.4)
 * [ ] $CFGS/f/flex/flex-2.6.4.eb (module: flex/2.6.4)
 * [ ] $CFGS/b/binutils/binutils-2.31.1.eb (module: binutils/2.31.1)
 * [ ] $CFGS/g/GCCcore/GCCcore-8.2.0.eb (module: GCCcore/8.2.0)
== Temporary log file(s) /tmp/eb-aTAklR/easybuild-dHqsp3.log* have been removed.
== Temporary directory /tmp/eb-aTAklR has been removed.
```

```
eb ~/tmp/easybuild_M4/prefix/software/EasyBuild/3.8.0/lib/python2.7/site-packages
/easybuild_easyconfigs-3.8.0-py2.7.egg/easybuild/easyconfigs/b/binutils/binutils-2.31.1.eb  -k -D                            
== temporary log file in case of crash /tmp/eb-VnvhZ6/easybuild-bQ2_gE.log                                                   
Dry run: printing build status of easyconfigs and dependencies
CFGS=/home/rschade/tmp/easybuild_M4/prefix/software/EasyBuild/3.8.0/lib/python2.7/site-packages/easybuild_easyconfigs-3.8.0-py2.7.egg/easybuild/easyconfigs
 * [ ] $CFGS/m/M4/M4-1.4.18.eb (module: M4/1.4.18)
 * [ ] $CFGS/b/Bison/Bison-3.0.5.eb (module: Bison/3.0.5)
 * [ ] $CFGS/z/zlib/zlib-1.2.11.eb (module: zlib/1.2.11)
 * [ ] $CFGS/h/help2man/help2man-1.47.4.eb (module: help2man/1.47.4)                                                         
 * [ ] $CFGS/m/M4/M4-1.4.17.eb (module: M4/1.4.17)
 * [ ] $CFGS/b/Bison/Bison-3.0.4.eb (module: Bison/3.0.4)
 * [ ] $CFGS/f/flex/flex-2.6.4.eb (module: flex/2.6.4)
 * [ ] $CFGS/b/binutils/binutils-2.31.1.eb (module: binutils/2.31.1)                                                         
== Temporary log file(s) /tmp/eb-VnvhZ6/easybuild-bQ2_gE.log* have been removed.                                             
== Temporary directory /tmp/eb-VnvhZ6 has been removed.
```

```
eb ~/tmp/easybuild_M4/prefix/software/EasyBuild/3.8.0/lib/python2.7/site-packages/easybuild_easyconfigs-3.8.0-py2.7.egg/easybuild/easyconfigs/f/flex/flex-2.6.4.eb  -k -D
== temporary log file in case of crash /tmp/eb-h81j31/easybuild-SItP2o.log
Dry run: printing build status of easyconfigs and dependencies
CFGS=/home/rschade/tmp/easybuild_M4/prefix/software/EasyBuild/3.8.0/lib/python2.7/site-packages/easybuild_easyconfigs-3.8.0-py2.7.egg/easybuild/easyconfigs
 * [ ] $CFGS/m/M4/M4-1.4.18.eb (module: M4/1.4.18)
 * [ ] $CFGS/h/help2man/help2man-1.47.4.eb (module: help2man/1.47.4)
 * [ ] $CFGS/m/M4/M4-1.4.17.eb (module: M4/1.4.17)
 * [ ] $CFGS/b/Bison/Bison-3.0.4.eb (module: Bison/3.0.4)
 * [ ] $CFGS/f/flex/flex-2.6.4.eb (module: flex/2.6.4)
== Temporary log file(s) /tmp/eb-h81j31/easybuild-SItP2o.log* have been removed.
== Temporary directory /tmp/eb-h81j31 has been removed.
```

```
eb ~/tmp/easybuild_M4/prefix/software/EasyBuild/3.8.0/lib/python2.7/site-packages/easybuild_easyconfigs-3.8.0-py2.7.egg/easybuild/easyconfigs/b/Bison/Bison-3.0.4.eb   -k -D
== temporary log file in case of crash /tmp/eb-bDMpgc/easybuild-3gq4dA.log
Dry run: printing build status of easyconfigs and dependencies
CFGS=/home/rschade/tmp/easybuild_M4/prefix/software/EasyBuild/3.8.0/lib/python2.7/site-packages/easybuild_easyconfigs-3.8.0-py2.7.egg/easybuild/easyconfigs
 * [ ] $CFGS/m/M4/M4-1.4.17.eb (module: M4/1.4.17)
 * [ ] $CFGS/b/Bison/Bison-3.0.4.eb (module: Bison/3.0.4)
== Temporary log file(s) /tmp/eb-bDMpgc/easybuild-3gq4dA.log* have been removed.
== Temporary directory /tmp/eb-bDMpgc has been removed.
```

Reply:
@robertschade It's indeed a bit silly that `flex-2.6.4.eb` depends on `M4-1.4.18` and (indirectly, via `Bison`) also on `M4-1.4.17`.

Since `Bison` is only a build dependency to `flex`, it's acceptable to make that change imho, but let's do it in a separate PR with sufficient motivation.

Reply:
I have created a PR with a lengthy explanation at https://github.com/easybuilders/easybuild-easyconfigs/pull/7414 for the change of the dependencies.

Reply:
Excellent, thank you very much for this @robertschade, let's follow up there w.r.t. the indirect `Bison` dep for `M4`.